internal/auth/oidc/service.go

```diff
@@ -2,88 +2,229 @@ package oidc
+	const op = "oidc.(Repository).oidcWrapper"
+	if k == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "missing kms")
+	}
+	if scopeId == "" {
+		return nil, errors.New(errors.InvalidParameter, op, "missing scope id")
+	}
+	if authMethodId == "" {
+		return nil, errors.New(errors.InvalidParameter, op, "missing auth method id")
+	}
+	opts := getOpts(opt...)
+	// get a specific oidcWrapper using the WithKeyId(...) option
+	oidcWrapper, err := k.GetWrapper(ctx, scopeId, kms.KeyPurposeOidc, kms.WithKeyId(opts.withKeyId))
+	if err != nil {
+		return nil, errors.Wrap(err, op, errors.WithMsg("unable to get oidc wrapper"))
+	}
+
+	// What derived key are we looking for?
+	keyId := derivedKeyId(derivedKeyPurposeState, oidcWrapper.KeyID(), authMethodId)
+	derivedWrapper, ok := k.GetDerivedPurposeCache().Load(keyId)
+	if ok {
+		return derivedWrapper.(*aead.Wrapper), nil
+	}
+
+	// okay, I guess we need to derive a new key for this combo of oidcWrapper and authMethodId
+	reader, err := kms.NewDerivedReader(oidcWrapper, 32, []byte(authMethodId), []byte(scopeId))
+	if err != nil {
+		return nil, errors.Wrap(err, op)
+	}
+	privKey, _, err := ed25519.GenerateKey(reader)
+	if err != nil {
+		return nil, errors.New(errors.Encrypt, op, "unable to generate key", errors.WithWrap(err))
+	}
+	wrapper := aead.NewWrapper(nil)
+	if _, err := wrapper.SetConfig(map[string]string{
+		"key_id": keyId,
+	}); err != nil {
+		return nil, errors.Wrap(err, op, errors.WithMsg(fmt.Sprintf("error setting config on aead wrapper in auth method %s", authMethodId)))
+	}
+	if err := wrapper.SetAESGCMKeyBytes(privKey); err != nil {
+		return nil, errors.Wrap(err, op, errors.WithMsg(fmt.Sprintf("error setting key bytes on aead wrapper in auth method %s", authMethodId)))
+	}
+	// store the derived key in our cache
+	k.GetDerivedPurposeCache().Store(keyId, wrapper)
+
+	return wrapper, nil
 }
 
-// TokenRequest is an oidc domain service function for processing a token
-// request from a Boundary client.  Token requests are the result of a Boundary
-// client polling the tokenUrl they received via StartAuth.  On success, it
-// returns Boundary token.
-//
-// * Decrypt the tokenRequestId.  If encryption fails, it returns an error.
+// derivedKeyId returns a key that represents the derived key
+func derivedKeyId(purpose derivedKeyPurpose, wrapperKeyId, authMethodId string) string {
+	return fmt.Sprintf("%s.%s.%s", purpose.String(), wrapperKeyId, authMethodId)
+}
+
+// derivedKeyPurpose represents the purpose of the derived key.
 //
-// * Use the authtoken.(Repository).IssueAuthToken to issue the request id's
-// token and mark it as issued in the repo.  If the token is already issue, an
-// error is returned.
-func TokenRequest(ctx context.Context, kms *kms.Kms, atRepoFn AuthTokenRepFactory, tokenRequestId string) (string, error) {
-	panic("to-do")
+// TODO (jimlambrt 3/2021) ask @jefferai if this should be moved to the kms package
```

Comment:
I'm fine with that if it's easier. I don't see any other needs _at the moment_ unless we refactor session establishment key derivation to use it (which we could do but I don't see a benefit right now). So also happy to have this stay here and we can unify later if needed.